### DISCUSSION TITLE
How to JAVA comment detection？

### DISCUSSION THREAD

**Discussion by JiayuK** [pmd/pmd#5298] [Q&A] [answered]
I want to write rules to detect comments on some variables in java code. May I ask what APIs I can use to get comments in the code?

**Comment by adangel** [comment, marked as answer]
Comments are not directly accessible within the AST (that's open for discussion/future enhancement to change this) - the comments are currently not represented as nodes. However, we have some utilities to access comments nevertheless.

E.g. the rule [CommentDefaultAccessModifier](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#commentdefaultaccessmodifier) makes use of [JavaComment.html#getLeadingComments(net.sourceforge.pmd.lang.java.ast.JavaNode)](https://docs.pmd-code.org/apidocs/pmd-java/7.7.0/net/sourceforge/pmd/lang/java/ast/JavaComment.html#getLeadingComments(net.sourceforge.pmd.lang.java.ast.JavaNode)) to retrieve the comments that appear just before a given node. Note, that the returned comment might also be a [JavadocComment](https://docs.pmd-code.org/apidocs/pmd-java/7.7.0/net/sourceforge/pmd/lang/java/ast/JavadocComment.html).

This also means, that comments are not accessible from XPath based rules, so you definitely need to write a [Java-based rule](https://docs.pmd-code.org/latest/pmd_userdocs_extending_writing_java_rules.html).

**Reply by JiayuK** [reply]
thx
